copy/copy.go

```diff
@@ -899,6 +899,12 @@ func (c *copier) copyBlobFromStream(ctx context.Context, srcStream io.Reader, sr
 		return types.BlobInfo{}, errors.Wrap(err, "Error writing blob")
 	}
 
+	// If we can modify the layer's blob, set the desired compression for it to be set in the manifest.
+	if canModifyBlob && !isConfig {
+		uploadedInfo.Compression = c.dest.DesiredLayerCompression()
+		uploadedInfo.CompressionAlgorithm = desiredCompressionFormat.Name()
+	}
```

Comment:
In the `else` case, is `Compression` is still set and/or consumed? I can’t see anything in here setting the values on every manifest read.

We either need the values to be always filled in and present (by every single producer), or there will need to be an “unknown”/“don’t edit” value.  (Abstractly, I’d weakly prefer the former, but that would probably benefit from wrapping `BlobInfo` in constructors to enforce the field presence. Adding a “don’t edit” state would probably be easier right now.)

Reply:
I think we can get the "don't edit" state by having a `Compression *compression.Algorithm` with `nil` being a "don't edit".

Reply:
Note that we need to differentiate between “decompress” and “don’t edit”. Would ”don’t edit” be (`LayerCompression.Compress`, `nil`)? That is good enough for a computer but rather confusing for people.

---

Actually the code already implements “don’t edit” using `PreserveOriginal`. That’s semantically… not _quite_ precise but close enough, and at least I can’t imagine what else the code should do with a `PreserveOriginal` value.

Reply:
> Actually the code already implements “don’t edit” using PreserveOriginal.

That's what I thought as well. I think the combination of operation and algorithm works well - although I think we shouldn't use a `compress.Algorithm` as it might trick users into believing they could specify custom (de)compression functions (we're only interested in the name).

Reply:
True, the `Compression` field is used in `Image.UpdatedImage`, which could be confusing to mean that the contents will get automatically recompressed, but the immediate container is called *`Manifest`*`UpdateOptions`…

Maybe name the field not `Compression`, but some smarter variant of `MIMETypeCompressionAdjustment`, then?

---

WRT the users passing in an unknown algorithm, that’s actually much less hassle with strings than with `compression.Algorithm` values (right now `Algorithm` is a private struct, so it’s completely impossible; even assuming it becomes a public interface, implementing an interface is more work. And maybe we can add a private function to the interface to make it exclusive, not sure whether Go allows that). And after all, your `UpdateLayerInfos` implementation already rejects unknown values, so I don’t think it matters either way.

---

This is basically the fallout of being lax before, and abusing `types.BlobInfo` as an universal carrier of various subsets of data instead of being explicit in the type system about what is / isn’t available/expected in which situations. Maybe there should be a `BlobInfo` and `BlobInfoEdits`. Or move the compression edits to be a standalone array in `ManifestUpdateInformation`?  

In particular it’s inconsistent / it does not make all that much sense to have _both_ `MediaType` and the two compression _edits_ to `MediaType` in a single structure… and then have `ManifestUpdateOptions` say that “`MediaType` fields are ignored”.

Yet fixing this cleanly has ABI implications, which immediately leads to thinking about which parts of the write path should be an ABI / “unstable public API” / …, and I suspect there isn’t much appetite for solving that this week.

Reply:
We're getting close! I will reflect on Monday about that. Thanks for sharing your thoughts in such great detail!

Reply:
I left it as is (for now). As you mention above, the compression function of `compression.Algorithm` is private, so there's no real risk at the moment.